fieldpath/set.go

```diff
@@ -206,6 +206,29 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
 	return subset
 }
 
+// Leaves returns a set containing only the leaf paths
+// of a set.
+func (s *Set) Leaves() *Set {
+	members := PathElementSet{}
+	for _, member := range s.Members.members {
+		isChild := false
+		for _, child := range s.Children.members {
```

Comment:
@lavalamp Do you have an opinion about the O(n^2) loop here? There's no significant different in benchmarks, but I'm wondering why.

Reply:
Can you binary search s.Children.members instead of scanning? Looks like they are [ordered](https://github.com/kubernetes-sigs/structured-merge-diff/blob/a4e00e9840c22df6cc47c60dcf65984228558387/fieldpath/set.go#L218
). I doubt it will matter for the typical case, but maybe help for the worst case?

Reply:
I.e. just do `s.Children.members.Descend(child.pathElement)`

Reply:
I can't quite tell if Descend is the right thing, but search is as easy as something like this
```
loc := sort.Search(len(s.members), func(i int) bool {
		return !s.members[i].pathElement.Less(pe)
	})
```

Reply:
But even that is `O(n * log p)` with `n` the number of members and `p` the number of children. Moving the two pointers should be `O(n + p)`

Reply:
This results in an almost identical benchmark to both the nested for loop and O(n) implementation from [here](https://github.com/kevindelgado/structured-merge-diff/blob/double-pointer-leaves/fieldpath/set.go#L209-L246).

You can see the code I tested this with [here](https://github.com/kevindelgado/structured-merge-diff/blob/binary-search-leaves/fieldpath/set.go#L209-L231) (in case it's not optimal or what you expected).

The (almost) identical benchmark is:
```
BenchmarkFieldSet/leaves-20-16                            319598              3893 ns/op            2250 B/op         41 allocs/op
BenchmarkFieldSet/leaves-50-16                             81356             14551 ns/op            6851 B/op        140 allocs/op
BenchmarkFieldSet/leaves-100-16                            21072             58362 ns/op           29149 B/op        632 allocs/op
BenchmarkFieldSet/leaves-500-16                             3243            365927 ns/op          146657 B/op       3215 allocs/op
BenchmarkFieldSet/leaves-1000-16                            1417            844380 ns/op          325139 B/op       7190 allocs/op
```

It seems like the random set generator might not ever be testing the worst case, unless there is some under the hood optimizations going on that make the O(n), O(n log n) and O(n^2) approaches all actually do the same thing.

Reply:
In summary,

benchmark results:
O(n) == O(n log n) == O(n^2)

so I went with O(n^2) because for readability because:
O(n^2) > [O(n log n)](https://github.com/kevindelgado/structured-merge-diff/blob/binary-search-leaves/fieldpath/set.go#L209-L231) > [O(n)](https://github.com/kevindelgado/structured-merge-diff/blob/double-pointer-leaves/fieldpath/set.go#L209-L246)

I can look into why the random set generation used by the benchmark test is not testing the worst case, but if we're short on time and need to land this asap, we might be better just picking one for now and optimizing later?

Reply:
@jpbetz @apelisse thoughts^

Reply:
I'll defer to @lavalamp on this.

Reply:
This file uses the double pointer logic about everywhere (and for good reasons), so I'd still go with that if possible. I've tried to write it, it's not that bad.

Reply:
done, ptal @apelisse